code/code/misc/tweaks.cc

```diff
@@ -0,0 +1,268 @@
+//////////////////////////////////////////////////////////////////////////
+//
+// SneezyMUD - All rights reserved, SneezyMUD Coding Team
+//
+//  tweaks.cc
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include <cmath>
+#include "tweaks.h"
+#include "database.h"
+#include "being.h"
+
+
+tweakEntry::tweakEntry(int i,double v, double t, double r) :
+  id(i),
+  cvalue(v),
+  tvalue(t),
+  rate(r)
+{
+}
+
+tweakEntry::~tweakEntry()
+{
+}
+
+tweakEntry *tweakInfoT::operator[] (const tweakTypeT t)
+{
+  return tweaks.at(t);
+}
+
+tweakTypeT & operator++(tweakTypeT &c, int)
+{
+  return c = (c == MAX_TWEAK_TYPES) ? TWEAK_NONE : tweakTypeT(c+1);
+}
+
+tweakInfoT::~tweakInfoT()
+{
+}
+
+tweakInfoT::tweakInfoT(): 
+    tweaks(MAX_TWEAK_TYPES + 1, nullptr), 
+    loaded(false)
+{
+  loaded=false;
+}
+
+int tweakInfoT::loadTweaks()
+{
+  TDatabase db(DB_SNEEZY);
+
+  //On load fill out all tweaks with 1.0.  This will make adding new tweaks easier by eliminating the need to write a migration.
+  for(tweakTypeT i=TWEAK_NONE;i<MAX_TWEAK_TYPES;i++){
+    tweaks[i] = new tweakEntry(0,1.0, 1.0, 0.0);
+  }
+
+  //the globaltweaks table contains multiple rows per tweak type
+  //We only want to keep the last tweak update of each tweak type in memory, so we do a greatest n per group query
+  db.query("select tweak_id, tweak_type, tweak_value, tweak_target, tweak_rate from globaltweaks a "
+           "where tweak_id = (select max( tweak_id ) from globaltweaks b where a.tweak_type = b.tweak_type)"
```

Comment:
I didn't get that part. Why multiple rows per tweakable?

Reply:
A new row is added every time the tweakable is changed by hand so the history can be shown.  I didn't want to add a new row every second in procTweakRates, so for tweakables that have rates set, there is one row to mark the manual set point and a row to hold the current value.